tests/Stub/TestLogger.php

```diff
@@ -0,0 +1,194 @@
+<?php
+
+namespace Psr\Log\Util\Tests\Stub;
+
+use Psr\Log\AbstractLogger;
+use Psr\Log\InvalidArgumentException;
+use ReflectionClass;
+
+/**
+ * Used for testing purposes.
+ *
+ * It records all records and gives you access to them for verification.
+ *
+ * @method bool hasEmergency($record)
+ * @method bool hasAlert($record)
+ * @method bool hasCritical($record)
+ * @method bool hasError($record)
+ * @method bool hasWarning($record)
+ * @method bool hasNotice($record)
+ * @method bool hasInfo($record)
+ * @method bool hasDebug($record)
+ *
+ * @method bool hasEmergencyRecords()
+ * @method bool hasAlertRecords()
+ * @method bool hasCriticalRecords()
+ * @method bool hasErrorRecords()
+ * @method bool hasWarningRecords()
+ * @method bool hasNoticeRecords()
+ * @method bool hasInfoRecords()
+ * @method bool hasDebugRecords()
+ *
+ * @method bool hasEmergencyThatContains($message)
+ * @method bool hasAlertThatContains($message)
+ * @method bool hasCriticalThatContains($message)
+ * @method bool hasErrorThatContains($message)
+ * @method bool hasWarningThatContains($message)
+ * @method bool hasNoticeThatContains($message)
+ * @method bool hasInfoThatContains($message)
+ * @method bool hasDebugThatContains($message)
+ *
+ * @method bool hasEmergencyThatMatches($message)
+ * @method bool hasAlertThatMatches($message)
+ * @method bool hasCriticalThatMatches($message)
+ * @method bool hasErrorThatMatches($message)
+ * @method bool hasWarningThatMatches($message)
+ * @method bool hasNoticeThatMatches($message)
+ * @method bool hasInfoThatMatches($message)
+ * @method bool hasDebugThatMatches($message)
+ *
+ * @method bool hasEmergencyThatPasses($message)
+ * @method bool hasAlertThatPasses($message)
+ * @method bool hasCriticalThatPasses($message)
+ * @method bool hasErrorThatPasses($message)
+ * @method bool hasWarningThatPasses($message)
+ * @method bool hasNoticeThatPasses($message)
+ * @method bool hasInfoThatPasses($message)
+ * @method bool hasDebugThatPasses($message)
+ */
+class TestLogger extends AbstractLogger
```

Comment:
I'll be honest, this whole class seems heavily over-engineered to me.  I'm working on the logger in TYPO3 right now, and using this as a test anon class:

```php
$this->trackingLogger = new class() implements LoggerInterface {
    use LoggerTrait;
    public array $records = [];
    public function log($level, $message, array $context = []): void
    {
        $this->records[] = [
            'level' => $level,
            'message' => $message,
            'context' => $context
        ];
    }
};
```

Having a pre-made class to help with that is well and good and I'm all for it, but I don't see why that needs 200 lines and dozens of virtual methods (especially when virtual methods are generally evil).

Reply:
IMO the PHP 5.3 requirement here is at play too, so we can ditch this all with a 2.0 release immediately afterward.

Reply:
Wouldn't just a `records(string $level): array` method be ample, and then I can test that however I want myself with the standard PHPUnit tools?

Reply:
Even in ancient PHP versions, `log()` like above and a `records(string $level): array` (even untyped) that returns an empty array if nothing was recorded would be sufficient.  Give or take context values, I suppose, but this whole thing can be vastly simplified.

Reply:
I completely agree with this class being over-engineered. In fact, I have had the thought quite a few times that some methods seem to be extra - and the only method e.g. `getRecords()` that is really necessary was missing - I added it now. Also, I'm not sure how or whether this was being used in the old package, because AFAIK the only test that is actually running now was added by me, and I can't really think about a different application for this class other than for being used in a test to prove that the test tests what it's suppose to test (sorry for the tautology).

Yet still, my humble opinion is that this change (the possible simplification of this class) should happen in the next BC-breaking release.